File: src/api/RecipeCosting.Api/Models/Entities/Ingredient.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using RecipeCosting.Api.Models.Enums;

namespace RecipeCosting.Api.Models.Entities;

/// <summary>
/// Something the baker buys, priced by the package it comes in.
/// </summary>
[Table("ingredients")]
public class Ingredient
{
    /// <summary>Identifier of the ingredient.</summary>
    [Key]
    [Column("id")]
    public int Id { get; set; }

    /// <summary>Name shown to the baker, such as "Condensed milk".</summary>
    [Required]
    [MaxLength(120)]
    [Column("name")]
    public string Name { get; set; } = string.Empty;

    /// <summary>How much the package holds, in the unit below.</summary>
    [Column("package_size", TypeName = "numeric(12,2)")]
    public decimal PackageSize { get; set; }

    /// <summary>The unit the package is measured in.</summary>
    [Column("package_unit")]
    public PackageUnit PackageUnit { get; set; }

    /// <summary>What the whole package costs.</summary>
    [Column("package_price", TypeName = "numeric(12,2)")]
    public decimal PackagePrice { get; set; }
}

File: src/web/src/app/core/ingredient.ts
export type PackageUnit = 'g' | 'ml' | 'un';

export interface UnitCost {
  amount: number;
  label: string;
}

export interface Ingredient {
  id: number;
  name: string;
  packageSize: number;
  packageUnit: PackageUnit;
  packagePrice: number;
  unitCost: UnitCost;
}

export interface IngredientRequest {
  name: string;
  packageSize: number;
  packageUnit: PackageUnit;
  packagePrice: number;
}

export const NO_UNIT_COST: UnitCost = { amount: 0, label: '' };
